File: ios-sdk/Sources/MxLSdk/Interaction/UserInteractionTracker.swift
import Foundation
import UIKit

/**
 * User interaction tracker for capturing user gestures and screen views.
 */
internal class UserInteractionTracker {
    private static var isInitialized = false
    private static var trackedViewControllers = Set<String>()
    
    /**
     * Initialize user interaction tracking.
     */
    static func initialize(configuration: SdkConfiguration) {
        guard configuration.enableUserTracking, !isInitialized else {
            return
        }
        
        isInitialized = true
        Logger.shared.info("User interaction tracking initialized")
    }
    
    /**
     * Track screen view for a view controller.
     */
    static func trackScreenView(_ viewController: UIViewController) {
        let sdkState = MxLSdk.getSdkState()
        guard let sdkState = sdkState, sdkState.configuration.enableUserTracking else {
            return
        }
        
        let viewControllerName = String(describing: type(of: viewController))
        guard !trackedViewControllers.contains(viewControllerName) else {
            return
        }
        
        trackedViewControllers.insert(viewControllerName)
        
        let screenViewEvent = ScreenViewEvent(
            sessionId: sdkState.sessionManager.getSessionId() ?? "unknown",
            userId: sdkState.sessionManager.getUserId(),
            screenName: viewControllerName,
            screenClass: String(describing: type(of: viewController)),
            timestamp: Date().timeIntervalSince1970 * 1000
        )
        
        storeScreenView(screenViewEvent, sdkState: sdkState)
    }
    
    /**
     * Track tap event.
     */
    static func trackTap(view: UIView, location: CGPoint) {
        let sdkState = MxLSdk.getSdkState()
        guard let sdkState = sdkState, sdkState.configuration.enableUserTracking else {
            return
        }
        
        let viewClass = String(describing: type(of: view))
        let viewTag = view.tag
        
        let tapEvent = TapEvent(
            sessionId: sdkState.sessionManager.getSessionId() ?? "unknown",
            userId: sdkState.sessionManager.getUserId(),
            viewTag: viewTag != 0 ? String(viewTag) : nil,
            viewClass: viewClass,
            x: Double(location.x),
            y: Double(location.y),
            timestamp: Date().timeIntervalSince1970 * 1000
        )
        
        storeTapEvent(tapEvent, sdkState: sdkState)
    }
    
    /**
     * Track custom event.
     */
    static func trackEvent(_ eventName: String, properties: [String: Any] = [:]) {
        let sdkState = MxLSdk.getSdkState()
        guard let sdkState = sdkState, sdkState.configuration.enableUserTracking else {
            return
        }
        
        let customEvent = CustomEvent(
            sessionId: sdkState.sessionManager.getSessionId() ?? "unknown",
            userId: sdkState.sessionManager.getUserId(),
            eventName: eventName,
            properties: properties,
            timestamp: Date().timeIntervalSince1970 * 1000
        )
        
        storeCustomEvent(customEvent, sdkState: sdkState)
    }
    
    /**
     * Store screen view event.
     */
    private static func storeScreenView(_ event: ScreenViewEvent, sdkState: SdkState) {
        // TODO: Store in database
        Logger.shared.debug("Screen view: \(event.screenName)")
    }
    
    /**
     * Store tap event.
     */
    private static func storeTapEvent(_ event: TapEvent, sdkState: SdkState) {
        // TODO: Store in database
        Logger.shared.debug("Tap event: \(event.viewClass) at (\(event.x), \(event.y))")
    }
    
    /**
     * Store custom event.
     */
    private static func storeCustomEvent(_ event: CustomEvent, sdkState: SdkState) {
        // TODO: Store in database
        Logger.shared.debug("Custom event: \(event.eventName)")
    }
}

struct ScreenViewEvent {
    let sessionId: String
    let userId: String?
    let screenName: String
    let screenClass: String
    let timestamp: TimeInterval
}

struct TapEvent {
    let sessionId: String
    let userId: String?
    let viewTag: String?
    let viewClass: String
    let x: Double
    let y: Double
    let timestamp: TimeInterval
}

struct CustomEvent {
    let sessionId: String
    let userId: String?
    let eventName: String
    let properties: [String: Any]
    let timestamp: TimeInterval
}

File: android-sdk/sdk/src/main/java/com/adlcom/mxl/sdk/interaction/UserInteractionTracker.kt
package com.adlcom.mxl.sdk.interaction

import android.app.Activity
import android.view.MotionEvent
import android.view.View
import android.view.ViewGroup
import com.adlcom.mxl.sdk.config.SdkConfiguration
import com.adlcom.mxl.sdk.internal.SdkState
import com.adlcom.mxl.sdk.internal.getSdkState
import com.adlcom.mxl.sdk.logging.Logger
import java.lang.reflect.Method

/**
 * User interaction tracker for capturing user gestures and screen views.
 */
object UserInteractionTracker {
    private var isInitialized = false
    private val trackedActivities = mutableSetOf<String>()

    /**
     * Initialize user interaction tracking.
     */
    fun initialize(configuration: SdkConfiguration) {
        if (!configuration.enableUserTracking || isInitialized) {
            return
        }

        isInitialized = true
        Logger.d("User interaction tracking initialized")
    }

    /**
     * Track screen view for an activity.
     */
    fun trackScreenView(activity: Activity) {
        val sdkState = getSdkState() ?: return
        if (!sdkState.configuration.enableUserTracking) {
            return
        }

        val activityName = activity.javaClass.simpleName
        if (trackedActivities.contains(activityName)) {
            return
        }

        trackedActivities.add(activityName)

        val screenViewEvent = ScreenViewEvent(
            sessionId = sdkState.sessionManager.getSessionId() ?: "unknown",
            userId = sdkState.sessionManager.getUserId(),
            screenName = activityName,
            screenClass = activity.javaClass.name,
            timestamp = System.currentTimeMillis()
        )

        storeScreenView(screenViewEvent, sdkState)
    }

    /**
     * Track tap event.
     */
    fun trackTap(view: View, event: MotionEvent) {
        val sdkState = getSdkState() ?: return
        if (!sdkState.configuration.enableUserTracking) {
            return
        }

        val viewId = view.id
        val viewTag = view.tag?.toString() ?: ""
        val viewClass = view.javaClass.simpleName

        val tapEvent = TapEvent(
            sessionId = sdkState.sessionManager.getSessionId() ?: "unknown",
            userId = sdkState.sessionManager.getUserId(),
            viewId = if (viewId != View.NO_ID) view.resources.getResourceEntryName(viewId) else null,
            viewTag = viewTag,
            viewClass = viewClass,
            x = event.x,
            y = event.y,
            timestamp = System.currentTimeMillis()
        )

        storeTapEvent(tapEvent, sdkState)
    }

    /**
     * Track custom event.
     */
    fun trackEvent(eventName: String, properties: Map<String, Any> = emptyMap()) {
        val sdkState = getSdkState() ?: return
        if (!sdkState.configuration.enableUserTracking) {
            return
        }

        val customEvent = CustomEvent(
            sessionId = sdkState.sessionManager.getSessionId() ?: "unknown",
            userId = sdkState.sessionManager.getUserId(),
            eventName = eventName,
            properties = properties,
            timestamp = System.currentTimeMillis()
        )

        storeCustomEvent(customEvent, sdkState)
    }

    /**
     * Store screen view event.
     */
    private fun storeScreenView(event: ScreenViewEvent, sdkState: SdkState) {
        kotlinx.coroutines.CoroutineScope(kotlinx.coroutines.Dispatchers.IO).launch {
            try {
                sdkState.storageManager.storeEvent(
                    eventType = "interaction",
                    eventData = event,
                    sessionId = event.sessionId,
                    userId = event.userId
                )
                Logger.d("Screen view stored: ${event.screenName}")
            } catch (e: Exception) {
                Logger.e("Error storing screen view event", e)
            }
        }
    }

    /**
     * Store tap event.
     */
    private fun storeTapEvent(event: TapEvent, sdkState: SdkState) {
        kotlinx.coroutines.CoroutineScope(kotlinx.coroutines.Dispatchers.IO).launch {
            try {
                sdkState.storageManager.storeEvent(
                    eventType = "interaction",
                    eventData = event,
                    sessionId = event.sessionId,
                    userId = event.userId
                )
                Logger.d("Tap event stored: ${event.viewClass} at (${event.x}, ${event.y})")
            } catch (e: Exception) {
                Logger.e("Error storing tap event", e)
            }
        }
    }

    /**
     * Store custom event.
     */
    private fun storeCustomEvent(event: CustomEvent, sdkState: SdkState) {
        kotlinx.coroutines.CoroutineScope(kotlinx.coroutines.Dispatchers.IO).launch {
            try {
                sdkState.storageManager.storeEvent(
                    eventType = "interaction",
                    eventData = event,
                    sessionId = event.sessionId,
                    userId = event.userId
                )
                Logger.d("Custom event stored: ${event.eventName}")
            } catch (e: Exception) {
                Logger.e("Error storing custom event", e)
            }
        }
    }
}

data class ScreenViewEvent(
    val sessionId: String,
    val userId: String?,
    val screenName: String,
    val screenClass: String,
    val timestamp: Long
)

data class TapEvent(
    val sessionId: String,
    val userId: String?,
    val viewId: String?,
    val viewTag: String,
    val viewClass: String,
    val x: Float,
    val y: Float,
    val timestamp: Long
)

data class CustomEvent(
    val sessionId: String,
    val userId: String?,
    val eventName: String,
    val properties: Map<String, Any>,
    val timestamp: Long
)

